File: hashtable.py
MaxSlots = 3
MaxBuckets = 30
TableSize = 20
OverflowSize = 10


class Bucket:

    def __init__(self):
        self.count = 0
        self.overflow = -1
        self.slots = {}


class HashTable:

    def __init__(self):
        self.table = []
        for i in range(0, MaxBuckets):
            self.table.append(Bucket())

    def ReadFile(self):
        f = open('./datain.dat', 'r')
        for i in f.readlines():
            self.Insert(self.Hash(i[:10]) % TableSize, i[10:])

    def Hash(self, key):
        return ord(key[2]) + ord(key[4]) + ord(key[6]) % TableSize

    def Insert(self, key, value):
        item = self.CollissionCheck(key)
        if item is not None:
            # collission is true
            if 2 in item.keys():
                index = self.CheckAvailableBuckets(key, 19, MaxBuckets)
                # overflow is true
                length = self.GetTableLength(key, index)
                while length == 3:
                    # overflow of overflow is true
                    index = self.CheckAvailableBuckets(key, index + 1,
                        MaxBuckets)
                    length = self.GetTableLength(key, index)
                if length > 0:
                    # insert into 'old' overflow bucket
                    self.table[index].slots[key][length] = value
                else:
                    # new overflow bucket
                    self.table[index].slots[key] = {}
                    self.table[index].slots[key][0] = value
            else:
                length = len(item)
                item[length] = value
                # overflow is false, collission still true
        else:
            #brand new key
            index = self.CheckAvailableBuckets(key, 0, TableSize)
            self.table[index].slots[key] = {}
            self.table[index].slots[key][0] = value

    def GetTableLength(self, key, index):
        if key in self.table[index].slots:
            length = len(self.table[index].slots[key])
        else:
            length = len(self.table[index].slots)
        return length

    def CollissionCheck(self, key):
        for item in self.table:
            if key in item.slots:
                return item.slots[key]
        return None

    def CheckAvailableBuckets(self, key, minimum, maximum):
        for i in range(minimum, maximum):
            if key in self.table[i].slots.keys():
                return i
            if len(self.table[i].slots.keys()) is 0:
                return i
        # Should never get here
        raise Exception('Ran out of overflow buckets!')


ht = HashTable()

ht.ReadFile()

File: assign1/hashtable.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>

enum SlotSize
{
	MaxSlots = 3
};

enum BucketSize
{
	MaxBuckets = 30
};

const int TableSize = 20;

class Slot
{	
	public:
		Slot()
		{
			// this->key = ;
			// this->value = NULL;
		}

		std::string key;
		std::string value;
};

class Bucket
{
	public:
		Bucket() 
		{
			this->count = 0;
			this->overflow = -1;
			this->slots.resize(MaxSlots);

		}

		short int count;
		short int overflow;
		std::vector<Slot> slots;
};

class HashTable
{
	public:
		HashTable()
		{
			for(int i = 0; i < MaxBuckets; ++i)
			{
				table.resize(i+1);
				Bucket *b = new Bucket();
				table.push_back(*b);
				delete b;
			}
			this->DataIn();
		}

		std::vector<Bucket> table;

		void DataIn()
		{
			std::string key;
			std::string value;
			std::ifstream file ("datain.dat");
			if(file.is_open())
			{
				int i = 0;
				while(file.good())
				{
					getline(file, key);
					value = key.substr(10, 20);
					key.resize(10);
					// this->table.at(i).slots.at(0).key = this->Hash(key);
					// std::cout << this->Hash(key) << std::endl;
					// this->table.at(i).slots.at(0).value = key.substr(10, 20);
					// std::cout << "Value: " << value << std::endl;
					// std::cout << i << std::endl;
					++i;
				}
			}
			// int hash_index = int(key[2]) + int(key[4]) + int(key[6]);
		}

		void Insert(std::string key, std::string value)
		{
			int count = NULL;
			int index = this->CollissionCheck(key);
			bool overflow = false;
			std::cout << "Index is: " << index << std::endl;
			if(index != 0)
			{
				//collission is true
				if(table.at(index).count >= 3)
				{
					//over flow is true
					int of_index = this->CheckAvailableOverflow(key);
					count = this->table.at(of_index).count++;
					this->table.at(of_index).slots.at(count).key = this->Hash(key);
					this->table.at(of_index).slots.at(count).value = value;
					std::cout << "HEre";
					this->table.at(index).overflow = of_index;							
				}
				else 
				{
					count = this->table.at(index).count++;
					this->table.at(index).slots.at(count).key = key;
					this->table.at(index).slots.at(count).value = value;

				}							

			}
			else
			{
				// no collission
				// fresh key
				index = CheckAvailablePrimary();
				this->table.at(index).slots.at(0).key = this->Hash(key);
				std::cout << this->table.at(index).slots.at(0).key << std::endl;
				std::cout << this->Hash(key);
				this->table.at(index).slots.at(0).value = value;
				this->table.at(index).count++;
			}								
		}

		int CheckAvailablePrimary()
		{
			for(std::vector<int>::size_type i = 0; i != this->table.size(); ++i)
			{
				if(this->table.at(i).count == 0)
				{
					// std::cout << "i is: " << i << std::endl;
					return i;
				}
			}

			return 0;
		}

		int CheckAvailableOverflow(std::string key)
		{
			for(std::vector<int>::size_type i = 10; i != this->table.size(); ++i)
			{
				for(std::vector<int>::size_type j = 0; j != this->table.at(i).slots.size(); ++j)
				{
					if(this->table.at(i).slots.at(0).key == key)
					{
						std::cout << "i is: " << i << std::endl;
						return i;
					}
				}
			}

			return 10;
		}

		int CollissionCheck(std::string key)
		{
			for(std::vector<int>::size_type i = 0; i != this->table.size(); ++i)
			{
				for(std::vector<int>::size_type j = 0; j != this->table.at(i).slots.size(); ++j)
				{
					// std::cout << "test: " << this->table.at(i).slots.at(0).key << std::endl;
					// std::cout << "test2: " << key <<
					if(this->table.at(i).slots.at(0).key == key)
					{
						std::cout << "i is: " << i << std::endl;
		
						return i;
					}
				}
			}

			return 0;
		}
	// private:
		int Hash(std::string key)
		{
			return (int(key[2]) + int(key[4]) + int(key[6])) % TableSize;
		}
};

int main(void)
{
	HashTable ht;
	ht.Insert("Thsadbaradfoj", "Jones");
	ht.Insert("Thsijsdfoj", "Jones");
	ht.Insert("Thsijsdfoj", "Jones");
	ht.Insert("Thsijsdfoj", "Jones");
	ht.Insert("Thsijsdfoj", "Jones");
	ht.Insert("Thsijsdfoj", "Jones");
	ht.Insert("Thsijsdfoj", "Jones");
	ht.Insert("Thsijsdfoj", "Jones");
	// std::string test = 5;
	// std::cout << test;
	//std::cout << ht.table.front().count;
	// ht.Insert("Bob");
	// std::cout << std::endl << hash_index << std::endl;
	std::vector<Bucket> table;
	return 0;
}

// Hash Table is just an array of buckets